### PR TITLE
HANDSHAKE_DONE

### PR DESCRIPTION
**Pull request by martinthomson**
This is incomplete in the sense that it doesn't discard handshake state (yet).  @ddragana, do you think that you could use this in combination with your previous code (and tests)?

### CODE REVIEW

File: neqo-transport/src/connection.rs [hunk 1]
```diff
@@ -2781,11 +2861,36 @@ mod tests {
         assert!(client.events().any(stream_readable));
     }
 
+    /// Getting the client and server to reach an idle state is surprisingly hard.
+    /// The server sends HANDSHAKE_DONE at the end of the handshake, and the client
+    /// doesn't immediately acknowledge it.
+
+    /// Force the client to send an ACK by having the server send two packets out
```
Comment (ddragana) [inline neqo-transport/src/connection.rs]: you can force it to send ack by adding some time to now() ACK_DALAY=200ms.

Comment (ddragana) [reply neqo-transport/src/connection.rs]: but you will mix times a bit. reordering packets is better.

**Comment by ddragana** [comment]
> This is incomplete in the sense that it doesn't discard handshake state (yet). @ddragana, do you think that you could use this in combination with your previous code (and tests)?

yes, this is fine. thanks!

**Comment by martinthomson** [comment]
This should be ready for you to take over then.  Or you can merge it and create a follow-up.  I want to do the packet changes I talked about for Retry.

**Comment by martinthomson** [comment]
I'd like to land this, so that I can use it as the basis of other changes.  Is there any reason why I couldn't?

File: neqo-transport/src/connection.rs [hunk 2]
```diff
@@ -271,6 +291,48 @@ impl IdleTimeout {
...
+enum StateSignaling {
+    Idle,
+    HandshakeDone,
+    ConnectionClose,
+    CloseSent,
+}
+
+impl StateSignaling {
+    pub fn handshake_done(&mut self) {
+        *self = Self::HandshakeDone
```
Comment (ddragana) [inline neqo-transport/src/connection.rs]: Can you add a debug assertion that we are in the right state? Also handle not transitioning from ConnectionClose and CloseSent into HandshakeDone for non-debug build.

File: neqo-transport/src/connection.rs [hunk 3]
```diff
@@ -271,6 +291,48 @@ impl IdleTimeout {
...
+    pub fn closing(&self) -> bool {
+        *self == Self::ConnectionClose
+    }
+
+    pub fn close(&mut self) {
+        *self = Self::ConnectionClose
+    }
+
+    pub fn close_sent(&mut self) {
+        *self = Self::CloseSent
```
Comment (ddragana) [inline neqo-transport/src/connection.rs]: add an assertion here as well.

**Comment by ddragana** [comment]
We can also land this and I will add assertion in my PR for a keys discarding.

**Pull request merged**